torch/csrc/utils/tensor_new.cpp

```diff
@@ -580,49 +548,70 @@ Tensor legacy_tensor_new(c10::DispatchKey dispatch_key, at::ScalarType scalar_ty
       // unless the sequences is a torch.Size
-      return legacy_new_from_sequence(dispatch_key, scalar_type, deviceOptional, r.pyobject(0));
+      return legacy_new_from_sequence(options, scalar_type, deviceOptional, r.pyobject(0));
     }
-    return new_with_sizes(dispatch_key, scalar_type, r.deviceOptional(1), r.intlist(0));
+    return new_with_sizes(options, scalar_type, r.deviceOptional(1), r.intlist(0));
   } else if (r.idx == 5) {
     auto deviceOptional = r.deviceOptional(1);
     check_legacy_ctor_device(dispatch_key, deviceOptional);
-    return legacy_new_from_sequence(dispatch_key, scalar_type, r.deviceOptional(1), r.pyobject(0));
+    return legacy_new_from_sequence(options, scalar_type, r.deviceOptional(1), r.pyobject(0));
   }
   throw std::runtime_error("new(): invalid arguments");
 }
 
 Tensor indexing_tensor_from_data(
-    c10::DispatchKey dispatch_key,
+    c10::TensorOptions options,
     at::ScalarType scalar_type,
     c10::optional<Device> device,
     PyObject* data) {
   // Specific to tensor indexing, converts an indexing list to an
   // indexing tensor (type Byte or Long)
   ScalarType inferred_scalar_type = infer_scalar_type(data);
   if (inferred_scalar_type == ScalarType::Byte || inferred_scalar_type == ScalarType::Bool) {
-    return internal_new_from_data(dispatch_key, inferred_scalar_type, std::move(device), data,
+    return internal_new_from_data(options, inferred_scalar_type, device, data,
                                   /*copy_variables=*/false, /*copy_numpy=*/false,
                                   /*type_inference=*/false);
   } else {
-    return internal_new_from_data(dispatch_key, scalar_type, std::move(device), data,
+    return internal_new_from_data(options, scalar_type, device, data,
                                   /*copy_variables=*/false, /*copy_numpy=*/false,
                                   /*type_inference=*/false);
   }
 }
 
+// Note [Ensuring sparse values and indices match devices]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// In all places where we construct indices, we read out options from values
+// (rather than use inferred_options).  Why?  This handles the case when
+// values is a CUDA tensor, but indices is a non-Tensor value (and the device
+// argument is not set).  Example:
+//
+//  torch.sparse_coo_tensor(([0, 1],), self.empty(2, 0).cuda(), (4, 0))
+//
+// Sparse tensors require both indices and values to live on the same device.
+// If values lives on CUDA, we can infer where the indices should live, and
+// should accept even ordinary index sequences (and just make sure we write them
+// into the correct device).  values is the ONLY way we know that the index
+// tensor should go to CUDA, so we have to get the information in somehow.
+//
+// This code is kind of jank.  For one, the dtype in options is silently ignored
+// by internal_new_from_data.  Also, in classic janky code style, it used to
```

Comment:
This is a choice we're making in adopting the new scheme though, right? The relevant subset of stuff carried by an options argument will be context-dependent.

Reply:
Not... really? Previously the code accepted DispatchKey, but ignored everything besides the Device associated with that DispatchKey. Options is a bigger type, but the mismatch in size was always there in the code.